api/document/tree.py

```diff
@@ -8,11 +8,66 @@
 
 PrimitiveDict = Dict[str, Any]
 
+T = TypeVar('T', bound='DocCursor')
```

Comment:
Note that we're using a [type variable with an upper bound](http://mypy.readthedocs.io/en/latest/generics.html#type-variable-upper-bound) here to ensure that our `DocCursor` methods that return instances of `DocCursor` are actually typed to the sub-class of the `DocCursor` they come from.  Otherwise e.g. `JSONAwareCursor.new_tree()` will actually return a `DocCursor` rather than a `JSONAwareCursor`.

Reply:
I know type vars are usually represented by single letters, but it makes the method type annotations hard to read, in my opinion. Would it be a bad idea to do something like
```python
DocCursorSubType = TypeVar('DocCursorSubType', bound='DocCursor')
```
?

Reply:
Yeah, I like that better than `T`.  Although I cry at how this will affect our 80-character line limit, since all methods using this type will need to repeat `DocCursorSubType` twice 😣.

Reply:
That said, though, I'd also argue that since the `TypeVar` *does* appear twice on method signatures, I think it actually compensates for the opaque-ness of the type variable name, as readers can actually just infer from the `self: T` or `cls: Type[T]` that it's referring to the parent class; they don't even really need to find the definition of `T`, in some sense. How about we leave it as-is and change it if folks find it confusing?